target/linux/ath79/dts/qca9531_comfast_cf-ew72.dts

```diff
@@ -0,0 +1,129 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+#include "qca953x.dtsi"
+
+/ {
+	compatible = "comfast,cf-ew72", "qca,qca9531";
+	model = "COMFAST CF-EW72";
+
+	aliases {
+		serial0 = &uart;
+		led-boot = &wan;
+		led-failsafe = &wan;
+		led-upgrade = &wan;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&jtag_disable_pins>;
+
+		lan: lan {
+			label = "cf-ew72:blue:lan";
+			gpios = <&gpio 2 GPIO_ACTIVE_LOW>;
+		};
+
+		wan: wan {
+			label = "cf-ew72:blue:wan";
+			gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
+		};
+
+		wlan {
+			label = "cf-ew72:blue:wlan";
+			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
+			linux,default-trigger = "phy0tpt";
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys"; // -polled
+		// poll-interval = <20>;
+
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
+			debounce-interval = <60>;
+		};
+	};
+
+	watchdog {
+		compatible = "linux,wdt-gpio";
+
+		gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
+		hw_algo = "toggle";
+		hw_margin_ms = <1200>;
+		always-running;
+	};
+};
+
+&pcie0 {
+	status = "okay";
+};
+
+&spi {
+	status = "okay";
+	num-cs = <1>;
+
+	flash@0 {
+		compatible = "winbond,w25q128", "jedec,spi-nor" ;
+		reg = <0>;
+		spi-max-frequency = <25000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			uboot: partition@0 {
+				label = "u-boot";
+				reg = <0x000000 0x010000>;
+				read-only;
+			};
+
+			art: partition@10000 {
+				label = "art";
+				reg = <0x010000 0x010000>;
+				read-only;
+			};
+
+			firmware: partition@20000 {
+				compatible = "denx,uimage";
+				label = "firmware";
+				reg = <0x020000 0xfd0000>;
+			};
+
+			nvram: partition@ff0000 {
+				label = "nvram";
+				reg = <0xff0000 0x010000>;
+				read-only;
+			};
+		};
+	};
+};
+
+&uart {
+	status = "okay";
+};
+
+&eth0 {
+	status = "okay";
+	phy-handle = <&swphy4>;
+	mtd-mac-address = <&art 0x0>;
+};
+
+&eth1 {
+	status = "okay";
+	mtd-mac-address = <&art 0x1002>;
+};
+
+&wmac {
+	status = "okay";
+	mtd-cal-data = <&art 0x1000>;
+	mtd-mac-address = <&art 0x6>;
```

Comment:
That's unprobable. Can you please provide a list of the MAC addresses on flash (art 0x0, 0x6, 0x1002, ...) and a list of the addresses which are used on stock FW (lan, wan, 2.4 GHz, ...)

For further reading: https://openwrt.org/docs/guide-developer/mac.address

Reply:
20:0d:b0:75:9b:f2 (0x0)
20:0d:b0:75:9b:f4 (0x6)
20:0d:b0:75:9b:f3 (0x1002)

These values match the stock firmware.  Do you need a dump of the ART partition?

Reply:
Well, it looks strange to me that you would "overwrite" the MAC address in the WiFi's calibration data (0x1002) with another one (0x6), but then use the WiFi address on eth1 ...

Reply:
Should I just remove the mtd-mac-address = <&art 0x6>; line?

Reply:
That depends on which of those the two MAC addresses matches the 2.4G's address on stock firmware...

Reply:
I'll attempt to reflash factory tonight and confirm.

Reply:
Confirmed the LAN port is F2, the WAN port is F3, the 5G radio is F4, and the 2.4G radio is F5 on the factory firmware.

Reply:
And in this firmware it looks like the 5G is F5 and the 2.4G is F4... weird.

Is there a way I can ensure the order is swapped, or does it matter for the WiFi interface (the WAN and LAN are correct)?  Sorry, still new to all this.

Reply:
So, let's collect our data:

```
use   stock    location
lan    f2         0x0
wan  f3          0x1002
5g     f4          0x6
2g     f5          0x5006
```

So, unless you mixed something up, I suspect the vendor just doesn't care what's stored in the calibration data and just increments the MAC address for its devices.

An based on this mess (I hesitate to use 5G calibration data to assign 2.4 G address, I'd recommend to just calculate all addresses based on an offset from 0x0, so

lan 0x0
wan 0x0 +1
...

Reply:
Does the device have a label with a MAC address on it?

Reply:
Yes, but only for the LAN port.  There is no label for WAN or for the 2G/5G radios.

Reply:
Okay, that's what I expected.

Reply:
So should I just define the LAN as-is, WAN as lan +1, and use the calibration data (remove the mtd-mac-address) line?

Reply:
And please add a nicer version of my improvised table to the commit message after you have squashed your changes.

Reply:
no, the calibration data would yield the wrong addresses.

Reply:
Okay.  eth0 in the devicetree (wan port, eth1 in openwrt):
        phy-handle = <&swphy4>;
        mtd-mac-address = <&art 0x0>;
        mtd-mac-address-increment = <1>;

eth1 in the devicetree (lan port, eth0 in openwrt):
        mtd-mac-address = <&art 0x0>;

wmac in the devicetree (2.4g in openwrt):
        mtd-mac-address = <&art 0x0>;
        mtd-mac-address-increment = <3>;

And coalescing cf-ew72 into the same 11-ath10k-caldata section as cf-e560ac gives me the exact same MAC addresses on Openwrt as they do on stock.  Now just to figure out the LEDs...